# Initial EDA and Workplan for Interconnect's churn prediction model

Interconnect, a leading telecom operator, seeks to proactively address customer churn by forecasting which clients are most likely to leave their services. To achieve this, the company aims to leverage advanced machine learning techniques to identify the key factors driving customer attrition and predict which customers are at the highest risk of churning.

The objective of this project is to develop and thoroughly evaluate a machine learning model that can accurately predict customer churn. This model will empower Interconnect to implement targeted retention strategies, minimize churn rates, and enhance overall customer satisfaction.

### Overview:
Data Collection: The dataset consists of customer data from the telecom operator Interconnect. This includes information about customer demographics, contract details, payment methods, and usage of various telecom services. The data is collected from Interconnect’s operational systems and reflects the behavior and characteristics of both churning and non-churning customers.

Preprocessing: The data will be preprocessed to prepare it for model training. This involves encoding categorical variables, scaling numerical features, and creating new features such as customer tenure and a binary churn indicator. The data will also be balanced to address any class imbalance between churning and non-churning customers.

Model Building: We will explore and build various machine learning models, including Logistic Regression, Decision Trees, Random Forest, Gradient Boosting Machines, and more advanced models like XGBoost and Neural Networks. The models will be trained to predict whether a customer will churn based on the available features.

Model Evaluation: The model's performance will be evaluated using metrics AUC-ROC, and accuracy. We will also analyze feature importance to understand which factors contribute most to customer churn. Ensuring the model is unbiased and generalizes well across different customer segments will be a priority.

Project Objective: The primary goal of this project is to build and evaluate a robust machine learning model that can accurately predict which customers are likely to churn. This model will enable Interconnect to take proactive measures, such as targeted promotions, to retain customers and reduce churn rates, ultimately contributing to better customer satisfaction and business sustainability.

## Initialization

In [2]:
import pandas as pd

## Load Data

In [3]:
contract_df = pd.read_csv('./datasets/contract.csv')
internet_df = pd.read_csv('./datasets/internet.csv')
personal_df = pd.read_csv('./datasets/personal.csv')
phone_df = pd.read_csv('./datasets/phone.csv')

## EDA

### Contract database

In [4]:
contract_df

customerID   BeginDate              EndDate            Type  \
0     7590-VHVEG  2020-01-01                   No  Month-to-month   
1     5575-GNVDE  2017-04-01                   No        One year   
2     3668-QPYBK  2019-10-01  2019-12-01 00:00:00  Month-to-month   
3     7795-CFOCW  2016-05-01                   No        One year   
4     9237-HQITU  2019-09-01  2019-11-01 00:00:00  Month-to-month   
...          ...         ...                  ...             ...   
7038  6840-RESVB  2018-02-01                   No        One year   
7039  2234-XADUH  2014-02-01                   No        One year   
7040  4801-JZAZL  2019-03-01                   No  Month-to-month   
7041  8361-LTMKD  2019-07-01  2019-11-01 00:00:00  Month-to-month   
7042  3186-AJIEK  2014-08-01                   No        Two year   

     PaperlessBilling              PaymentMethod  MonthlyCharges TotalCharges  
0                 Yes           Electronic check           29.85        29.85  
1                  No               Mailed check           56.95       1889.5  
2                 Yes               Mailed check           53.85       108.15  
3                  No  Bank transfer (automatic)           42.30      1840.75  
4                 Yes           Electronic check           70.70       151.65  
...               ...                        ...             ...          ...  
7038              Yes               Mailed check           84.80       1990.5  
7039              Yes    Credit card (automatic)          103.20       7362.9  
7040              Yes           Electronic check           29.60       346.45  
7041              Yes               Mailed check           74.40        306.6  
7042              Yes  Bank transfer (automatic)          105.65       6844.5  

[7043 rows x 8 columns]

In [5]:
contract_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   BeginDate         7043 non-null   object 
 2   EndDate           7043 non-null   object 
 3   Type              7043 non-null   object 
 4   PaperlessBilling  7043 non-null   object 
 5   PaymentMethod     7043 non-null   object 
 6   MonthlyCharges    7043 non-null   float64
 7   TotalCharges      7043 non-null   object 
dtypes: float64(1), object(7)
memory usage: 440.3+ KB


In [6]:
contract_df.isna().sum()

customerID          0
BeginDate           0
EndDate             0
Type                0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
dtype: int64

In [7]:
contract_df.duplicated().sum()

0

In [8]:
contract_df['EndDate'].value_counts()

EndDate
No                     5174
2019-11-01 00:00:00     485
2019-12-01 00:00:00     466
2020-01-01 00:00:00     460
2019-10-01 00:00:00     458
Name: count, dtype: int64

In [9]:
contract_df['PaymentMethod'].value_counts()

PaymentMethod
Electronic check             2365
Mailed check                 1612
Bank transfer (automatic)    1544
Credit card (automatic)      1522
Name: count, dtype: int64

In [33]:
contract_df[contract_df['EndDate'] != 'No']['Type'].value_counts()

Type
Month-to-month    1655
One year           166
Two year            48
Name: count, dtype: int64

In [34]:
contract_df[contract_df['EndDate'] != 'No']['PaymentMethod'].value_counts()

PaymentMethod
Electronic check             1071
Mailed check                  308
Bank transfer (automatic)     258
Credit card (automatic)       232
Name: count, dtype: int64

In [39]:
len(contract_df[contract_df['PaymentMethod'] == 'Electronic check'][contract_df['EndDate'] != 'No']) / len(contract_df[contract_df['EndDate'] != 'No'])

C:\Users\crisl\AppData\Local\Temp\ipykernel_16520\2105670291.py:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  len(contract_df[contract_df['PaymentMethod'] == 'Electronic check'][contract_df['EndDate'] != 'No']) / len(contract_df[contract_df['EndDate'] != 'No'])


0.5730337078651685

In [47]:
len(contract_df[contract_df['EndDate'] != 'No'][(contract_df['PaymentMethod'] == 'Bank transfer (automatic)') | (contract_df['PaymentMethod'] == 'Credit card (automatic)')]) / len(contract_df[contract_df['EndDate'] != "No"])

C:\Users\crisl\AppData\Local\Temp\ipykernel_16520\360942023.py:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  len(contract_df[contract_df['EndDate'] != 'No'][(contract_df['PaymentMethod'] == 'Bank transfer (automatic)') | (contract_df['PaymentMethod'] == 'Credit card (automatic)')]) / len(contract_df[contract_df['EndDate'] != "No"])


0.26217228464419473

In [10]:
len(contract_df[contract_df['EndDate'] == 'No']) / len(contract_df['EndDate'])

0.7346301292063041

In [11]:
contract_df['BeginDate'] = pd.to_datetime(contract_df['BeginDate'])

min_date = contract_df['BeginDate'].min()
max_date = contract_df['BeginDate'].max()

print("Minimum Date:", min_date)
print("Maximum Date:", max_date)

Minimum Date: 2013-10-01 00:00:00
Maximum Date: 2020-02-01 00:00:00


### Notes about the Contract Database:
- 73% of users are still with us.
- All users that have churned seem to have left after 2018.
- 57% of users that have churned had electronic checks as their payment method, compared to 26% of users that churned who had automatic payments set up. The remaining 17% of churned users used mailed checks as their method.
- No missing or duplicate values/rows.

### Preprocessing Tasks for Contract Database:
- Column names need to be corrected using lowercase letters and underscores.
- BeginDate, and EndDate values and column needs to be changed to datetime.
- Hour of day will be removed from EndDate, as the values are all 00:00:00.
- Yes/No values will be changed to 1/0's, Dataype will be changed accordingly.

### Feature Engineering for Contract Database:
- Type, PaperlessBilling, PaymentMethod will all be encoded for model.
- Create a Churn indicator based on the EndDate.
- Calculate the duration of each contract ContractLength as the difference between BeginDate and EndDate.
- Generate feature based on ContractType*MonthlyCharges.

### Questions regarding Contract Database:
- Are there any price discounts based on PaymentMethod/PaperlessBilling/Type? (Ex: Discount for those that use Automatic payments)
- How are monthly payments influenced by having multiple services? (Ex: Internet and Phone)
- All users that have churned seem to have left after 2018, is this data complete?
- What effect does paperless billing have on the customer? Is there a effective difference when it comes to price?

###  Internet Database

In [12]:
internet_df

customerID InternetService OnlineSecurity OnlineBackup DeviceProtection  \
0     7590-VHVEG             DSL             No          Yes               No   
1     5575-GNVDE             DSL            Yes           No              Yes   
2     3668-QPYBK             DSL            Yes          Yes               No   
3     7795-CFOCW             DSL            Yes           No              Yes   
4     9237-HQITU     Fiber optic             No           No               No   
...          ...             ...            ...          ...              ...   
5512  6840-RESVB             DSL            Yes           No              Yes   
5513  2234-XADUH     Fiber optic             No          Yes              Yes   
5514  4801-JZAZL             DSL            Yes           No               No   
5515  8361-LTMKD     Fiber optic             No           No               No   
5516  3186-AJIEK     Fiber optic            Yes           No              Yes   

     TechSupport StreamingTV StreamingMovies  
0             No          No              No  
1             No          No              No  
2             No          No              No  
3            Yes          No              No  
4             No          No              No  
...          ...         ...             ...  
5512         Yes         Yes             Yes  
5513          No         Yes             Yes  
5514          No          No              No  
5515          No          No              No  
5516         Yes         Yes             Yes  

[5517 rows x 8 columns]

In [13]:
internet_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5517 entries, 0 to 5516
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   customerID        5517 non-null   object
 1   InternetService   5517 non-null   object
 2   OnlineSecurity    5517 non-null   object
 3   OnlineBackup      5517 non-null   object
 4   DeviceProtection  5517 non-null   object
 5   TechSupport       5517 non-null   object
 6   StreamingTV       5517 non-null   object
 7   StreamingMovies   5517 non-null   object
dtypes: object(8)
memory usage: 344.9+ KB


In [14]:
internet_df.isna().sum()

customerID          0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
dtype: int64

In [15]:
internet_df.duplicated().sum()

0

In [26]:
internet_df['InternetService'].value_counts()

InternetService
Fiber optic    3096
DSL            2421
Name: count, dtype: int64

In [25]:
len(internet_df) / len(contract_df)

0.7833309669175067

In [49]:
len(internet_df[internet_df['InternetService'] == "Fiber optic"]) / len(internet_df)

0.5611745513866232

### Notes about the Internet Database:
- 78% of customers have an Internet Service Plan.
- 56% of internet users have fiber optic as their service.
- No missing or duplicate values/rows.

### Preprocessing Tasks for Contract Database:
- Column names need to be corrected using lowercase letters and underscores.
- Yes/No values will be changed to 1/0's, Dataype will be changed accordingly.
- Services may be consolidated based on answers to questions.

### Questions regarding Internet Database:
- Are OnlineSecurity/Device protection seperate services, or do they come together?
- Are TvStreaming/MovieStreaming protection seperate services, or do they come together?
- Are there any bundle discounts for multiple services?
- Is there user data available on how much internet data is used by customer?

### Phone Database

In [16]:
phone_df

customerID MultipleLines
0     5575-GNVDE            No
1     3668-QPYBK            No
2     9237-HQITU            No
3     9305-CDSKC           Yes
4     1452-KIOVK           Yes
...          ...           ...
6356  2569-WGERO            No
6357  6840-RESVB           Yes
6358  2234-XADUH           Yes
6359  8361-LTMKD           Yes
6360  3186-AJIEK            No

[6361 rows x 2 columns]

In [17]:
phone_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6361 entries, 0 to 6360
Data columns (total 2 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   customerID     6361 non-null   object
 1   MultipleLines  6361 non-null   object
dtypes: object(2)
memory usage: 99.5+ KB


In [18]:
phone_df.isna().sum()

customerID       0
MultipleLines    0
dtype: int64

In [19]:
phone_df.duplicated().sum()

0

In [20]:
len(phone_df) / len(contract_df)

0.9031662643759761

In [28]:
len(phone_df[phone_df['MultipleLines'] == 'Yes']) / len(phone_df)

0.46706492689828644

### Notes about the Phone Database:
- No missing or duplicate values/rows.
- 90% of customers have a Phone Service plan.
- 48% of customers have multiple lines.

### Preprocessing Tasks for Contract Database:
- Column names need to be corrected using lowercase letters and underscores.

### Questions regarding Phone Database:
- Is there a cost to having multiple lines? What is that cost?
- Because this is a landline phone, I imagine phone usage is practically unlimited. If there is a 'data' or usage cap, what is that?

### Personal Database

In [29]:
personal_df

customerID  gender  SeniorCitizen Partner Dependents
0     7590-VHVEG  Female              0     Yes         No
1     5575-GNVDE    Male              0      No         No
2     3668-QPYBK    Male              0      No         No
3     7795-CFOCW    Male              0      No         No
4     9237-HQITU  Female              0      No         No
...          ...     ...            ...     ...        ...
7038  6840-RESVB    Male              0     Yes        Yes
7039  2234-XADUH  Female              0     Yes        Yes
7040  4801-JZAZL  Female              0     Yes        Yes
7041  8361-LTMKD    Male              1     Yes         No
7042  3186-AJIEK    Male              0      No         No

[7043 rows x 5 columns]

In [30]:
personal_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   customerID     7043 non-null   object
 1   gender         7043 non-null   object
 2   SeniorCitizen  7043 non-null   int64 
 3   Partner        7043 non-null   object
 4   Dependents     7043 non-null   object
dtypes: int64(1), object(4)
memory usage: 275.2+ KB


In [51]:
personal_df.isna().sum()

customerID       0
gender           0
SeniorCitizen    0
Partner          0
Dependents       0
dtype: int64

In [52]:
personal_df.duplicated().sum()

0

In [29]:
len(personal_df[personal_df['Dependents'] == 'Yes']) / len(personal_df)

0.2995882436461735

In [30]:
len(personal_df[personal_df['Partner'] == 'Yes']) / len(personal_df)

0.4830327985233565

In [32]:
len(personal_df[personal_df['SeniorCitizen'] == 1]) / len(personal_df)

0.1621468124378816

### Notes about the Personal Database:
- 30% of customers have Dependents.
- 48% of customers have Partners.
- 16% of customers are Seniors.

### Preprocessing Tasks for Contract Database:
- Column names need to be corrected using lowercase letters and underscores.
- Yes/No and Male/Female values will be changed to 1/0's, Dataype will be changed accordingly.

### Questions regarding Personal Database:
- Are there any senior discounts?
- Are there price variables for internet data used? Different prices based on data speed? (Having partner/dependents may influence data usage.)

## Proposed Workplan

### Our goal is to build a model that will predict customers that will churn. In order to build that model and achieve the goal, the following steps will be performed:

- Load necessary Libraries
- Download and Read CSV files as dataframes
- Perform data preprocessing for the dataframes
    1. Column names will be changed as stated above.
    2. Datatypes will be changed as appropriate. (Yes/No will be changed into 1/0's)
    3. Columns may be removed, based on answers to questions and impact to churning. (Ex: Paperless Billing)    
    4. Feature engineering and data enrichment will be done as stated above, and depending on answers to questions asked. (Churn column will be created based on values for EndDate,)
    5. Dataframes will be merged on CustomerID into a main dataframe, with missing values due to merge being filled in.
- EDA will be performed.
    1. Data will be encoded using OH Encoding.
    2. Data will be scaled for uniformity.
    3. Classes will be balanced with upsampling/downsampling.
    4. Correlations between features will be examined and charted.
    5. Distribution of Churn will be examined and charted.
    6. Importace of features will be examined and visualized.
- Model Selection will be performed
    1. Data will be split into Training, Validation and Test Sets at a 3:1:1 ratio.
    2. Various models will be trained and examined, including: Logistic Regression, Decision Trees, Random Forest, Gradient Boosting Machines, XGBoost, LightGBM, and Neural Networks.
    3. GridSearch and K-Fold Cross-Validation will be applied to models in order to find the optimal Hyperparameters.
    4. Models will be evaluated based on AUC-ROC, and accuracy.
    5. Models will be analysed against the Test Set for final scores.
